src/libs/history.c

```diff
@@ -169,6 +175,67 @@ static GtkWidget *_lib_history_create_button(dt_lib_module_t *self, int num, con
   return widget;
 }
 
+static GList *_duplicate_history(GList *hist)
+{
+  GList *result = NULL;
+
+  GList *h = g_list_first(hist);
+  while(h)
+  {
+    const dt_dev_history_item_t *old = (dt_dev_history_item_t *)(h->data);
+
+    dt_dev_history_item_t *new = (dt_dev_history_item_t *)g_malloc(sizeof(dt_dev_history_item_t));
+
+    new->module = old->module;
+    new->enabled = old->enabled;
+    new->multi_priority = old->multi_priority;
+    snprintf(new->multi_name, sizeof(old->multi_name), "%s", old->multi_name);
+    new->focus_hash = old->focus_hash;
+
+    new->params = malloc(old->module->params_size);
+    new->blend_params = malloc(sizeof(dt_develop_blend_params_t));
+
+    memcpy(new->params, old->params, old->module->params_size);
+    memcpy(new->blend_params, old->blend_params, sizeof(dt_develop_blend_params_t));
+
+    result = g_list_append(result, new);
+
+    h = g_list_next(h);
+  }
+  return result;
+}
+
+static void pop_undo(dt_lib_module_t *self, dt_undo_type_t type, dt_undo_data_t *data)
+{
+  if(type == DT_UNDO_HISTORY)
+  {
+    dt_lib_history_t *d = (dt_lib_history_t *)self->data;
+    dt_undo_history_t *hist = (dt_undo_history_t *)data;
+
+    const GList *current_params = darktable.develop->history;
+    const int end = darktable.develop->history_end;
+
+    darktable.develop->history = hist->snapshot;
+    darktable.develop->history_end = hist->end;
+
+    hist->snapshot = current_params;
+    hist->end = end;
+
+    // disable recording undo as the _lib_history_change_callback will be triggered by the calls below
+    d->record_undo = FALSE;
+
+    dt_dev_write_history(darktable.develop);
+    dt_dev_reload_history_items(darktable.develop);
+  }
+}
+
+static void _history_undo_data_free(dt_undo_data_t *data)
+{
+  dt_undo_history_t *hist = (dt_undo_history_t *)data;
+  GList *snapshot = hist->snapshot;
+  g_list_free_full(snapshot, dt_dev_free_history_item);
```

Comment:
No it is freed in undo.c. See _free_undo_date. This routine is only for freeing any internal data if needed. It is not needed in map.c for example.